### PR TITLE
New `unreachable!` macro.

### LINKED ISSUE #4749: RFC: `unreachable!` method to express unreachable code explicitly
Sometimes we use `raise "BUG: ..."` to express unreachable code. For example:

```crystal
enum FooBar
  Foo
  Bar
end

def foo_bar_flip(foo_bar : FooBar)
  case foo_bar
  when .foo?
    Bar
  when .bar?
    Foo
  else
    raise "BUG: unreachable" # <- this!!
  end
end
```

This `raise "BUG: unreachable"` is important because it is required to infer `foo_bar_flip`'s result type as `FooBar` correctly, and if not, it is inferred to `FooBar?` due to the case of matching no `when`-clause (it is *unreachable* case also).

There are many unreachable code: `git grep 'raise "BUG: ' | wc -l` in crystal repository shows 81 in fact. So, I think stdlib should provide the method to express such an unreachable code more explicitly than `raise "BUG: "`. This method makes easier to read code. I think such a method is called `unreachable!`.

`unreachable!` usage:

```crystal
def foo_bar_flip(foo_bar : FooBar)
  case foo_bar
  when .foo?
    Bar
  when .bar?
    Foo
  else
    unreachable!
  end
end
```

And `unreachable!` spec:

  - `unreachable!` raises an error, so this type is `NoReturn`.
  - `unreachable!` has default message to an error, but we can specify more helpful message.

- - -

My questions:

  1. I think `unreachable!` is the best name for such a method. However we have some ideas: `unreach` (verb?), `unreachable` (without bang) and etc... Which do you think the best?

  2. Should stdlib have `unreachable!` specific error class? In other words, should `unreachable!` raise `UnreachableError`-like object? Or `Exception` object simply?

  3. What is the default message? I think it is `BUG: unreachable`. Or, should `unreachable` have the default message?

     * I think having the default message is important because *this is unreachable*, so we never see this message in normal cases. Do you take a time to consider such a message?

Related issues:

  - #1846: @asterite suggested such a method named `assert_empty_type` and @RX14 suggested `impossible!`.

And I was inspired by Rust's [`unreachable!` macro](https://doc.rust-lang.org/std/macro.unreachable.html). Thank you.

### PR DESCRIPTION
**Pull request by erdnaxeli**
This fixes #4749.

### CODE REVIEW

File: src/macros.cr [hunk 1]
```diff
@@ -185,3 +185,27 @@ macro p!(*exps)
...
 end
+
+# Raises an exception.
+# Useful to signal an unreachable section of the code.
+#
+# ```
+# values = ["a", "b"]
+# idx = Random.rand(2)
+#
+# case values[idx]
```
Comment (straight-shoota) [inline src/macros.cr:196]: This is a bit shorter and similarly expressive:
```suggestion
# case values.sample
```

**Review by straight-shoota — APPROVED** [review]
It doesn't seem this really needs to be a macro. Making `unreachable!` a method should be fine. It doesn't do anything that needs to happen at compile time.

I'm not sure why [Rust's implementation](https://doc.rust-lang.org/std/macro.unreachable.html) is a macro. But I also don't know how Rust macros work...

**Comment by asterite** [comment]
In the original issue the problem was that you needed to raise if an enum value wasn't covered. That's not the case anymore if you use "case in".

There's really no reason to introduce something that's essentially a very simple raise. And raise is more explicit. People will have to learn about this cryptic unreachable! thing when it's just raising.

Let's not merge this.

**Comment by erdnaxeli** [comment]
> In the original issue the problem was that you needed to raise if an enum value wasn't covered. That's not the case anymore if you use "case in".
> 
> There's really no reason to introduce something that's essentially a very simple raise. And raise is more explicit. People will have to learn about this cryptic unreachable! thing when it's just raising.
> 
> Let's not merge this.

It seems to me that the issue conclusion was exactly the opposite :sweat_smile: . But I understand. We can then close this PR and the linked issue too.

File: src/macros.cr (hunk 1 above)
Comment (erdnaxeli) [reply src/macros.cr:196]: I was looking for this exact method, but I didn't found it.

**Comment by asterite** [comment]
Yes, and that happened before we had exhaustive checks built into the language. And the general consensus didn't include my approval.

**Comment by erdnaxeli** [comment]
So let's close it.

**Pull request closed without merging**